waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,381 @@
+          current-time (t/now)
+          _ (when-not (t/before? current-time not-on-or-after)
+              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                              {:current-time current-time
+                               :expiry-time not-on-or-after
+                               :status 400})))
+          auth-cookie-expiry-date (t/min-date not-on-or-after (t/plus current-time (t/days 1)))
+          auth-cookie-age-in-seconds (-> current-time
+                                       (t/interval auth-cookie-expiry-date)
+                                       t/in-seconds)
+          {:keys [authorization/principal authorization/user] :as auth-params-map}
+          (auth/auth-params-map saml-principal)]
+      (auth/handle-request-auth (constantly {:body ""
+                                             :headers {"location" redirect-url}
+                                             :status 303})
+                                request principal auth-params-map password auth-cookie-age-in-seconds))
+    (throw (ex-info "Missing saml-auth-data from SAML authenticated redirect message"
+                    {:status 400}))))
+
+(defn- escape-xml-string
+  "Escape a string for use in an XML document."
+  [str]
+  (string/escape str {\' "&apos;"
+                      \" "&quot;"
+                      \& "&amp;"
+                      \< "&lt;"
+                      \> "&gt;"}))
+
+(let [saml-authentication-request-template-fn
+      (template/fn
+        [{:keys [time-issued saml-service-name saml-id acs-url idp-uri]}]
+        (slurp (io/resource "auth/saml-authentication-request.xml")))]
+  (defn render-saml-authentication-request-template
+    "Render the SAML authentication request XML."
+    [context]
+    (saml-authentication-request-template-fn (pc/map-vals escape-xml-string context))))
+
+(defn create-request
+  "Return XML elements that represent a SAML 2.0 auth request."
+  [time-issued saml-service-name saml-id acs-url idp-uri]
+  (render-saml-authentication-request-template (utils/keys-map time-issued saml-service-name saml-id acs-url idp-uri)))
+
+(def instant-format (f/formatters :date-time-no-ms))
+
+(defn make-issue-instant
+  "Converts a date-time to a SAML 2.0 time string."
+  []
+  (du/date-to-str (t/now) instant-format))
+
+(defn create-request-factory!
+  "Creates new requests for a particular service, format, and acs-url."
+  [idp-uri saml-service-name acs-url]
+  ;;; Bootstrap opensaml when we create a request factory.
+  (DefaultBootstrap/bootstrap)
+  #(create-request (make-issue-instant)
+                   saml-service-name
+                   (str "WAITER-" (utils/unique-identifier))
+                   acs-url
+                   idp-uri))
+
+(defmacro endpoint-with-waiter-metrics
+  "Calls body, wrapping with timer, count, concurrent count, and rate metrics"
+  [classifier nested-path & body]
+  `(:out (metrics/with-timer
```

Comment:
We should extract `classifier` and `nested-path` into local variables, e.g. `classifier#` and `nested-path#`,  first since they are referenced multiple times.

Reply:
why is it not ok to reference multiple times, isn't the macro expanded once?

Reply:
It is call-by-name, so will re-evaluate the body every time it is referenced.